### PR TITLE
fix: close pr issues and stop running handlers twice

### LINKED ISSUE #710: Linked Pull Request Not Closed On Expired Task
This should be diagnosed and fixed. 

> @wannacfuture - Releasing the bounty back to dev pool because the allocated duration already ended! 
Last activity time: Fri Aug 18 2023 23:07:15 GMT+0000 (Coordinated Universal Time)

_Originally posted by @ubiquibot in https://github.com/ubiquity/ubiquibot/issues/431#issuecomment-1694037026_

### PR DESCRIPTION
**Pull request by seprintour**
Resolves #710 and #716 

QA: https://github.com/Seprintour-Test/test/issues/22

Steps: 
- `closePullRequestForAnIssue` was called twice, i removed it from the unassign function so unassigning is now certain (no blocking request, closing PRs by looping was pretty intensive)

- Refactored `closePullRequestForAnIssue` to use the gitIssueParser to fetch the exact PR open for an issue instead of looping all PRs. (Closing is more certain and 1 memory intensive task down)

### CODE REVIEW

**Comment by seprintour** [comment]
@rndquu  @whilefoo

File: src/handlers/assign/action.ts [hunk 1]
```diff
@@ -72,13 +73,19 @@ export const closePullRequestForAnIssue = async (): Promise<void> => {
   const payload = context.payload as Payload;
   if (!payload.issue?.number) return;
 
-  const prs = await getOpenedPullRequestsForAnIssue(payload.issue.number, "");
+  const prs = await gitIssueParser({
+    owner: payload.repository.owner.login,
+    repo: payload.repository.name,
+    issue_number: payload.issue.number,
+  });
+
```
Comment (wannacfuture) [inline src/handlers/assign/action.ts:81]: This requires organization & repo name check and so on. 

https://github.com/ubiquity/ubiquibot/pull/657

It is being updated as gitLinkedPrParser and gitLinkedIssueParser. You can wait till it is finished.

Comment (whilefoo) [reply src/handlers/assign/action.ts:81]: But `gitLinkedPrParser` will only return the latest merged PR, this feature should close all PRs related to the issue

Comment (seprintour) [reply src/handlers/assign/action.ts:81]: > This requires organization & repo name check and so on. 
> 
> https://github.com/ubiquity/ubiquibot/pull/657
> 
> It is being updated as gitLinkedPrParser and gitLinkedIssueParser. You can wait till it is finished.

I don't think that'd work for this, this gets multiple linked PR. I think we should move on with this

Comment (wannacfuture) [reply src/handlers/assign/action.ts:81]: You can just refer them(gitLinkedPrParser, and issueParser) and add some features to gitIssueParser - maybe org & repo name check will be required.

Comment (seprintour) [reply src/handlers/assign/action.ts:81]: > You can just refer them(gitLinkedPrParser, and issueParser) and add some features to gitIssueParser - maybe org & repo name check will be required.

Done

**Comment by seprintour** [comment]
Update here?

**Review by whilefoo — CHANGES_REQUESTED** [review]
can you rework `gitLinkedPrParser` to return an array of PRs and move other functionality (like retrieving PR and checking merged date to get latest merged PR) outside the function. basically you are merging your work into `gitLinkedPrParser` and then removing `gitIssueParser`

**Comment by seprintour** [comment]
> can you rework `gitLinkedPrParser` to return an array of PRs and move other functionality (like retrieving PR and checking merged date to get latest merged PR) outside the function. basically you are merging your work into `gitLinkedPrParser` and then removing `gitIssueParser`

I don't understand what you mean by get it outside the function

**Comment by whilefoo** [comment]
functions calling `gitLinkedPrParser` can make additional transformations and `gitLinkedPrParser` should only be focused on returning array of PRs to make it generic as possible

**Comment by seprintour** [comment]
> functions calling `gitLinkedPrParser` can make additional transformations and `gitLinkedPrParser` should only be focused on returning array of PRs to make it generic as possible

Done, check it out @whilefoo

File: src/helpers/parser.ts [hunk 2]
```diff
@@ -49,38 +40,69 @@ export const gitLinkedIssueParser = async ({ owner, repo, pull_number }: GitPars
...
+        const prHref = `https://github.com${prUrl}`;
+        prs.push({ number: prNumber, href: prHref });
       }
     }
-    return linkedPullRequest;
+    return latest ? linkedPullRequest : prs;
   } catch (error) {
     logger.error(`${JSON.stringify(error)}`);
-    return null;
+    return latest
```
Comment (whilefoo) [inline src/helpers/parser.ts:99]: it should return null in any case

File: src/helpers/parser.ts [hunk 3]
```diff
@@ -49,38 +40,69 @@ export const gitLinkedIssueParser = async ({ owner, repo, pull_number }: GitPars
...
-      if (Number.isNaN(Number(prNumber))) return null;
-      const pr = await getPullByNumber(context, Number(prNumber));
-      if (!pr || !pr.merged) continue;
 
-      if (!linkedPullRequest) linkedPullRequest = pr;
-      else if (linkedPullRequest.merged_at && pr.merged_at && new Date(linkedPullRequest.merged_at) < new Date(pr.merged_at)) {
-        linkedPullRequest = pr;
+      const prNumber = Number(parts[parts.length - 1]);
+
+      if (latest) {
```
Comment (whilefoo) [inline src/helpers/parser.ts:82]: like I said this functionality should be in calling functions, the only purpose of this function should be to get linked PRs

File: src/helpers/parser.ts [hunk 4]
```diff
@@ -49,38 +40,69 @@ export const gitLinkedIssueParser = async ({ owner, repo, pull_number }: GitPars
   }
 };
 
-export const gitLinkedPrParser = async ({ owner, repo, issue_number }: GitParser) => {
+export const gitLinkedPrParser = async ({ owner, repo, issue_number }: GitParser): Promise<HTMLElement[]> => {
```
Comment (whilefoo) [inline src/helpers/parser.ts:43]: why does it return html elements?

**Review by whilefoo — CHANGES_REQUESTED** [review]

File: src/helpers/parser.ts (hunk 4 above)
Comment (seprintour) [reply src/helpers/parser.ts:43]: Because it's parsed html 🤔

File: src/helpers/parser.ts (hunk 3 above)
Comment (seprintour) [reply src/helpers/parser.ts:82]: You can view the file in full to fully get it, there's the function you want to be generic  and there's another to get useful information from the parsed data

File: src/helpers/parser.ts (hunk 4 above)
Comment (whilefoo) [reply src/helpers/parser.ts:43]: `gitLinkedPrParser` and `getLinkedPrs` should be one function, I think there's no need to separate

File: src/helpers/parser.ts (hunk 3 above)
Comment (whilefoo) [reply src/helpers/parser.ts:82]: the function should parse html and return array of PRs, other things should be done by the calling function - no need for latest functionality in this function

File: src/helpers/parser.ts (hunk 4 above)
Comment (seprintour) [reply src/helpers/parser.ts:43]: > `gitLinkedPrParser` and `getLinkedPrs` should be one function, I think there's no need to separate

I don't understand anymore, you said you wanted `gitLinkedPrParser` generic, now it is and you want it to go back to how it was before.

I just separated it and made `gitLinkedPrParser` generic returning array of PRs and the other function handles the response and converts it to return either the latest PR or array of PR with full href ready to use

Comment (whilefoo) [reply src/helpers/parser.ts:43]: yes `gitLinkedPrParser` should be generic in a sense that it retrieves HTML, parses it, and returns an array of PRs. It doesn't need any other functionality. Everything else should be done by functions calling `gitLinkedPrParser`.

Currently `gitLinkedPrParser` doesn't return array of PRs but just HTML elements, we don't need just HTML anywhere in the code.

Comment (seprintour) [reply src/helpers/parser.ts]: > Currently `gitLinkedPrParser` doesn't return array of PRs but just HTML elements, we don't need just HTML anywhere in the code.

Please check it out now

Comment (whilefoo) [reply src/helpers/parser.ts]: looks mostly good, I'm sorry if there was miscommunication between us

File: src/helpers/parser.ts [hunk 5]
```diff
@@ -2,32 +2,24 @@ import axios from "axios";
...
 import { getBotContext, getLogger } from "../bindings";
-import { Payload } from "../types";
+import { Endpoints } from "@octokit/types";
 
 interface GitParser {
   owner: string;
   repo: string;
   issue_number?: number;
   pull_number?: number;
+  latest?: boolean;
```
Comment (whilefoo) [inline src/helpers/parser.ts:12]: this isn't used anymore, right?

File: src/helpers/parser.ts [hunk 6]
```diff
@@ -2,32 +2,24 @@ import axios from "axios";
...
-      return true;
-    } else {
-      //no LinkedPRs
-      return false;
-    }
-  } catch (error) {
-    return true;
-  }
-};
+export interface PRsParserResponse {
```
Comment (whilefoo) [inline src/helpers/parser.ts:15]:
```suggestion
export interface LinkedPR {
```

File: src/helpers/parser.ts [hunk 7]
```diff
@@ -49,38 +41,53 @@ export const gitLinkedIssueParser = async ({ owner, repo, pull_number }: GitPars
...
-      else if (linkedPullRequest.merged_at && pr.merged_at && new Date(linkedPullRequest.merged_at) < new Date(pr.merged_at)) {
-        linkedPullRequest = pr;
-      }
+      const prUrl = linkedPr.querySelector("a")?.attrs?.href;
+
+      if (!prUrl) continue;
+
+      const parts = prUrl.split("/");
+      // extract the organization name and repo name from the link:(e.g. "
+      const prOrganization = parts[parts.length - 4];
```
Comment (whilefoo) [inline src/helpers/parser.ts:61]: check if array size is at least 4

File: src/helpers/parser.ts (hunk 4 above)
Comment (seprintour) [reply src/helpers/parser.ts]: > looks mostly good, I'm sorry if there was miscommunication between us

Thanks @whilefoo, any other edits I have to make.

You used keyword *mostly*

File: src/helpers/parser.ts [hunk 8]
```diff
@@ -49,38 +41,53 @@ export const gitLinkedIssueParser = async ({ owner, repo, pull_number }: GitPars
...
+    return [];
   }
 };
+
+export const getLatestPullRequest = async (prs: PRsParserResponse[]): Promise<ListPullsByNumberResponse["data"] | null> => {
+  const context = getBotContext();
+  let linkedPullRequest = null;
+  for (const _pr of prs) {
+    if (Number.isNaN(_pr.prNumber)) return null;
+    const pr = (await getPullByNumber(context, _pr.prNumber)) as ListPullsByNumberResponse["data"];
```
Comment (whilefoo) [inline src/helpers/parser.ts:83]:
```suggestion
    const pr = await getPullByNumber(context, _pr.prNumber);
```
I think it's better to make it implicit

File: src/helpers/parser.ts [hunk 9]
```diff
@@ -49,38 +41,53 @@ export const gitLinkedIssueParser = async ({ owner, repo, pull_number }: GitPars
...
+
+    return prData;
   } catch (error) {
     logger.error(`${JSON.stringify(error)}`);
-    return null;
+    return [];
   }
 };
+
+export const getLatestPullRequest = async (prs: PRsParserResponse[]): Promise<ListPullsByNumberResponse["data"] | null> => {
```
Comment (whilefoo) [inline src/helpers/parser.ts:78]:
```suggestion
export const getLatestPullRequest = async (prs: PRsParserResponse[]) => {
```

File: src/helpers/parser.ts [hunk 10]
```diff
@@ -2,32 +2,24 @@ import axios from "axios";
...
-  }
-};
+export interface PRsParserResponse {
+  prOrganization: string;
+  prRepository: string;
+  prNumber: number;
+  prHref: string;
+}
+
+export type ListPullsByNumberResponse = Endpoints["GET /repos/{owner}/{repo}/pulls/{pull_number}"]["response"];
```
Comment (whilefoo) [inline src/helpers/parser.ts:22]:
```suggestion
```

**Review by whilefoo — CHANGES_REQUESTED** [review]

File: src/helpers/parser.ts (hunk 5 above)
Comment (seprintour) [reply src/helpers/parser.ts:12]: Yea, is no longer in use

File: src/helpers/parser.ts (hunk 7 above)
Comment (seprintour) [reply src/helpers/parser.ts:61]: > check if array size is at least 4

Done all

**Review by wannacfuture — APPROVED** [review]
Looks good to me

**Comment by seprintour** [comment]
> can you post QA?

I have QA in description

**Comment by seprintour** [comment]
@0xcodercrane ping

**Review by 0xcodercrane — APPROVED** [review]

**Pull request merged**